src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    /// <summary>
+    /// Default JsonTypeInfo resolver.
+    /// </summary>
+    internal sealed class DefaultJsonTypeInfoResolver : IJsonTypeInfoResolver
+    {
+        private bool _mutable = true;
+
+        /// <summary>
+        /// Constructs DefaultJsonTypeInfoResolver.
+        /// </summary>
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
+        public DefaultJsonTypeInfoResolver()
+        {
+            Modifiers = new ConfigurationList<Action<JsonTypeInfo>>() { VerifyMutable = VerifyMutable };
+        }
+
+        /// <inheritdoc/>
+        public JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
+        {
+            _mutable = false;
```

Comment:
This wouldn't prevent users from adding modifiers to the default options instance before any resolutions have taken place. Possible workarounds might be to only lock the instance used in `JsonSerializerOptions.Default` but it's going to be hairy. I'm starting to think that making the default resolver mutable is a desing mistake, and we should instead go back to using a combinator for subscribing callbacks.